core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -207,7 +210,23 @@ private[spark] class Executor(
 
         // Run the actual task and measure its runtime.
         taskStart = System.currentTimeMillis()
-        val value = task.run(taskAttemptId = taskId, attemptNumber = attemptNumber)
+        var succeeded: Boolean = false
+        val value = try {
+          val value = task.run(taskAttemptId = taskId, attemptNumber = attemptNumber)
+          succeeded = true
+          value
+        } finally {
+          // Release managed memory used by this task
+          val freedMemory = taskMemoryManager.cleanUpAllAllocatedMemory()
+          if (succeeded && freedMemory > 0) {
```

Comment:
I also agree that we should enforce this for failures as well.  In the worst case, I suppose our failure message is masked by a memory leak message, but that's fine for now as this is disabled by default and is only used in tests.